pkg/controller/attach_detach.go

```diff
@@ -151,48 +166,47 @@ func (r *ReconcileAttachDetach) triggerAttach(ctx context.Context, azVolumeAttac
 		return err
 	}
 
-	klog.Infof("Attaching volume (%s) to node (%s)", azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName)
-
-	// initiate goroutine to attach volume
+	workflow.Logger().Info("Attaching volume")
+	waitForGoroutine = true
 	//nolint:contextcheck // Attach is performed asynchronously by the reconciler; context is not inherited by design
 	go func() {
-		cloudCtx, cloudCancel := context.WithTimeout(context.Background(), cloudTimeout)
+		var attachErr error
+		defer func() { workflow.Finish(attachErr) }()
+		goCtx := workflow.SaveToContext(context.Background())
+		cloudCtx, cloudCancel := context.WithTimeout(goCtx, cloudTimeout)
 		defer cloudCancel()
-
 		// attempt to attach the disk to a node
-		response, attachErr := r.attachVolume(cloudCtx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Spec.VolumeContext)
+		var response map[string]string
+		response, attachErr = r.attachVolume(cloudCtx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Spec.VolumeContext)
 		// if the disk is attached to a different node
 		if danglingAttachErr, ok := attachErr.(*volerr.DanglingAttachError); ok {
 			// get disk, current node and attachment name
-			diskName, err := azureutils.GetDiskName(azVolumeAttachment.Spec.VolumeID)
-			if err != nil {
-				klog.Warningf("failed to extract disk name from volumeID (%s): %v", azVolumeAttachment.Spec.VolumeID, err)
-			}
 			currentNodeName := string(danglingAttachErr.CurrentNode)
-			currentAttachmentName := azureutils.GetAzVolumeAttachmentName(diskName, currentNodeName)
+			currentAttachmentName := azureutils.GetAzVolumeAttachmentName(azVolumeAttachment.Spec.VolumeName, currentNodeName)
+			workflow.Logger().Infof("Dangling attach detected for %s", currentNodeName)
 
 			// check if AzVolumeAttachment exists for the existing attachment
-			_, err = r.controllerSharedState.azClient.DiskV1beta1().AzVolumeAttachments(r.controllerSharedState.objectNamespace).Get(cloudCtx, currentAttachmentName, metav1.GetOptions{})
+			_, err := r.controllerSharedState.azClient.DiskV1beta1().AzVolumeAttachments(r.controllerSharedState.objectNamespace).Get(cloudCtx, currentAttachmentName, metav1.GetOptions{})
 			var detachErr error
 			if errors.IsNotFound(err) {
 				// AzVolumeAttachment doesn't exist so we only need to detach disk from cloud
-				detachErr = r.cloudDiskAttacher.UnpublishVolume(ctx, azVolumeAttachment.Spec.VolumeID, currentNodeName)
+				detachErr = r.cloudDiskAttacher.UnpublishVolume(goCtx, azVolumeAttachment.Spec.VolumeID, currentNodeName)
 				if detachErr != nil {
-					klog.Errorf("failed to detach volume (%s) from node (%s). Error: %v", azVolumeAttachment.Spec.VolumeID, currentNodeName, err)
+					workflow.Logger().Errorf(err, "failed to detach dangling volume (%s) from node (%s). Error: %v", azVolumeAttachment.Spec.VolumeID, currentNodeName, err)
```

Comment:
We should be logging `detachErr` here since we know `err` is "not found". Same with lines 205 & 209.